runtime/src/bank/builtins/core_bpf_migration/mod.rs

```diff
@@ -0,0 +1,121 @@
+mod bpf_upgradeable;
+mod builtin;
+pub(crate) mod error;
+
+use {
+    crate::bank::Bank,
+    bpf_upgradeable::BpfUpgradeableConfig,
+    builtin::BuiltinConfig,
+    error::CoreBpfMigrationError,
+    solana_sdk::{
+        account::{Account, AccountSharedData},
+        bpf_loader_upgradeable::{UpgradeableLoaderState, ID as BPF_LOADER_UPGRADEABLE_ID},
+        pubkey::Pubkey,
+    },
+    std::sync::atomic::Ordering::Relaxed,
+};
+
+/// Sets up a Core BPF migration for a built-in program.
+pub enum CoreBpfMigration {
+    Builtin,
+    Ephemeral,
+}
+
+/// Configurations for migrating a built-in program to Core BPF.
+pub struct CoreBpfMigrationConfig {
+    /// The source program ID to replace the builtin with.
+    pub source_program_id: Pubkey,
+    /// The feature gate to trigger the migration to Core BPF.
+    /// Note: This feature gate should never be the same as any builtin's
+    /// `enable_feature_id`. It should always be a feature gate that will be
+    /// activated after the builtin is already enabled.
```

Comment:
Yep, it would just be a Core BPF program that needs no migration, since it never existed as a builtin before.

I think in that spirit, as we either migrate programs or introduce new ones, it's definitely worth documenting these programs as Core BPF and what that means. Maybe in the specifications?

Reply:
Yeah that makes sense!